address/address.go

```diff
@@ -47,327 +68,268 @@ const (
-	if err != nil {
-		return errors.Wrap(err, "unable to decode")
-	}
-
-	if _, err := NetworkFromString(hrp); err != nil {
-		return errors.Wrap(err, "invalid network")
-	}
-
-	if version != Version {
-		return fmt.Errorf("invalid version: version=%d", version)
-	}
-
-	if len(data) != HashLength {
-		return fmt.Errorf("invalid data length: len=%d", len(data))
-	}
-
-	return nil
+// Empty returns true if the address is empty, false otherwise.
+func (a Address) Empty() bool {
+	return a == Undef
 }
 
-// Empty returns true if the address is empty.
-func (a Address) Empty() bool {
-	return a == (Address{})
+// Unmarshal unmarshals the cbor bytes into the address.
+func (a Address) Unmarshal(b []byte) error {
+	return cbor.DecodeInto(b, &a)
 }
 
-// Network returns the network of the address.
-func (a Address) Network() Network {
-	return a[0]
+// Marshal marshals the address to cbor.
+func (a Address) Marshal() ([]byte, error) {
+	return cbor.DumpObject(a)
 }
 
-// Version returns the version of the address.
-func (a Address) Version() byte {
-	return a[1]
+// UnmarshalJSON implements the json unmarshal interface.
+func (a *Address) UnmarshalJSON(b []byte) error {
+	in := strings.TrimSuffix(strings.TrimPrefix(string(b), `"`), `"`)
+	addr, err := decode(in)
+	if err != nil {
+		return err
+	}
+	*a = addr
+	return nil
 }
 
-// Hash returns the hash part of the address.
-func (a Address) Hash() []byte {
-	return a[2:]
+// MarshalJSON implements the json marshal interface.
+func (a Address) MarshalJSON() ([]byte, error) {
+	return json.Marshal(a.String())
 }
 
 // Format implements the Formatter interface.
 func (a Address) Format(f fmt.State, c rune) {
 	switch c {
 	case 'v':
-		fmt.Fprintf(f, "[%s - %x - %x]", NetworkToString(a.Network()), a.Version(), a.Hash()) // nolint: errcheck
+		fmt.Fprintf(f, "[%x - %x]", a.Protocol(), a.Payload()) // nolint: errcheck
 	case 's':
 		fmt.Fprintf(f, "%s", a.String()) // nolint: errcheck
 	default:
 		fmt.Fprintf(f, "%"+string(c), a.Bytes()) // nolint: errcheck
 	}
 }
 
-// MarshalText implements the TextMarshaler interface.
-func (a Address) MarshalText() ([]byte, error) {
-	if a == (Address{}) {
-		return nil, nil
-	}
-
-	return []byte(a.String()), nil
+// NewIDAddress returns an address using the ID protocol.
+func NewIDAddress(id uint64) (Address, error) {
+	return newAddress(ID, leb128.FromUInt64(id))
 }
 
-// UnmarshalText implements the TextUnmarshaler interface.
-func (a *Address) UnmarshalText(in []byte) error {
-	if len(in) == 0 {
-		return nil
-	}
+// NewSecp256k1Address returns an address using the SECP256K1 protocol.
+func NewSecp256k1Address(pubkey []byte) (Address, error) {
+	return newAddress(SECP256K1, addressHash(pubkey))
+}
 
-	out, err := NewFromString(string(in))
-	if err != nil {
-		return err
-	}
+// NewActorAddress returns an address using the Actor protocol.
+func NewActorAddress(data []byte) (Address, error) {
+	return newAddress(Actor, addressHash(data))
```

Comment:
check length of input

Reply:
length check is done in `newAddress`: https://github.com/filecoin-project/go-filecoin/pull/2186/files#diff-2026d132a1aa6ec72a1af72e5763011cR218.